src/execution_plan/optimizations/traverse_order.c

```diff
@@ -5,74 +5,229 @@
 */
 
 #include "./traverse_order.h"
+#include "../../util/arr.h"
 #include "../../util/vector.h"
+#include "../../util/rmalloc.h"
+#include <assert.h>
 
-/* Given a set of algebraic expressions and the entire filter tree,
- * suggest traversal entry point to be either the first expression or the
- * last expression, in the future we'll want to be able to begin traversal from 
- * any expression. */
-TRAVERSE_ORDER determineTraverseOrder(const FT_FilterNode *filterTree,
-                                      AlgebraicExpression **exps,
-                                      size_t expCount) {
+#define T 1     // Transpose penalty.
+#define F 4 * T // Filter score.
 
-    if(expCount == 1 || !filterTree) {
-        return TRAVERSE_ORDER_FIRST;
+typedef AlgebraicExpression** Arrangement;
+
+// Create a new arrangement.
+Arrangement Arrangement_New(uint size) {
+    return rm_malloc(sizeof(AlgebraicExpression*) * size);
+}
+
+// Clone arrangement.
+Arrangement Arrangement_Clone(const Arrangement arrangement, uint size) {
+    assert(arrangement);
+    Arrangement clone = Arrangement_New(size);
+    memcpy(clone, arrangement, sizeof(AlgebraicExpression*) * size);
+    return clone;
+}
+
+// Print arrangement.
+void Arrangement_Print(Arrangement arrangement, uint size) {
+    printf("Arrangement_Print\n");
+    for(uint i = 0; i < size; i++) {
+        AlgebraicExpression *exp = arrangement[i];
+        printf("%d, src: %s, dest: %s\n", i, exp->src_node->alias, exp->dest_node->alias);
     }
-    
-    char *destAlias;
-    char *srcAlias;
-    bool firstExpLabeled = false;
-    bool lastExpLabeled = false;
+}
+
+// Free arrangement.
+void Arrangement_Free(Arrangement arrangement) {
+    assert(arrangement);
+    rm_free(arrangement);
+}
+
+// Computes x!
+static uint factorial(uint x) {
+    uint res = 1;
```

Comment:
The largest factorial less than UINT_MAX is 12!, which (in addition to being just crazy to think about) makes me feel like we should use `ulong`s in this function.

Reply:
Crazy indeed.